proposals/2020-07-28_keyring-cmm/proposal.md

```diff
@@ -0,0 +1,293 @@
+  This proposal calls this the [Set Unspecified Request Values CMM](#set-unspecified-request-values-cmm).
+
+While the latter implementation might seem to be so simple it doesn't need to be its own CMM,
+it is necessary in order to be able to express the Default CMM as a specific composition
+while still keeping the Keyring CMM implementation simple and focused.
+
+## Drawbacks
+
+This proposal SHOULD NOT have any drawbacks.
+
+## Security Implications
+
+This proposal SHOULD NOT have any security implications.
+
+## Operational Implications
+
+This proposal SHOULD NOT have any operational implications.
+
+This proposal MUST maintain behavior for customers who are using a Default CMM.
+
+## Guide-level Explanation
+
+This proposal defines the [Keyring CMM](#keyring-cmm) and
+the [Master Key Provider CMM](#master-key-provider-cmm)
+as base CMM implementations that can be used "as is" or composed with other CMMs.
+
+This proposal also defines a [Set Unspecified Request Values CMM](#set-unspecified-request-values-cmm)
+as a CMM implementation that can wrap others in order to default the algorithm suite used to a specific value.
+
+This proposal also defines a [Default CMM](#default-cmm) as a specific composition of CMMs
+that the AWS Encryption SDK provides as a safe default.
+
+Customers SHOULD either use the [Default CMM](#default-cmm)
+or compose a CMM using the AWS Encryption SDK's provided CMM implementations
+(including the Keyring CMM).
+
+### Default CMM
+
+The Default CMM is a specific configuration of a
+CMM implementation provided by the AWS Encryption SDK that takes a keyring as input.
+
+The specific CMM configuration describes the set of CMM behaviors and configuration options
+that the AWS Encryption SDK recomends as a sensible default.
+
+The behavior we define as default is the behavior of the Default CMM prior to this proposal.
+
+As the AWS Encryption SDK provides more CMM implementations in the future that would benefit default use cases,
+we expect to update the Default CMM to a configuration which composes such CMMs together in order to provide
+useful properties.
+Any update to the composition defined by Default CMM MUST be backwards compatable.
+
+To ensure that its use is correct by construction,
+the instantiation of the Default CMM MUST NOT take any options
+except for the keyring that should be composed with the Default CMM.
+
+### Keyring CMM
+
+The Keyring CMM is a base CMM implementation that
+[gets encryption materials](../../framework/cmm-interface.md#get-encryption-materials)
+and [decrypts materials](../../framework/cmm-interface.md#decrypt-materials)
+through use of a [keyring](../../framework/keyring-interface.md) configured on instantiation.
+
+It requires an algorithm suite to be set in the Get Encryption Materials Request.
+
+Any desired CMM implementation that intends to use keyrings to ensure valid materials SHOULD
+compose with the Keyring CMM.
+
+### Set Unspecified Request Values CMM
```

Comment:
Could we have the Encrypt operation fill in the default when not provided as input, instead of it being the CMM's responsibility?

Reply:
We have considered this (and this is what Java/Python originally did, I believe), but we want this to be defined in the CMM for two reasons:
- The logic for everything the ESDK expresses as a "default" is encapsulated at the same level, in the Default CMM that we provide to the user. This is simple, and means that we have only one place in code where we define what "default" is.
- If this logic is in the CMM as opposed to the client level, then any other code that theoretically can be built on top of CMMs is able to grab the defaults, instead of needing to duplicate the "set some default for the algorithm suite" behavior.

Reply:
On thinking about this further, I'm not really convinced by either reason:

* "in the Encrypt operation" would be an equally valid single place to encapsulate the defaults
* Setting a default value doesn't seem to be difficult enough to worry about never duplicating IF there is significant code out there using CMMs outside the context of Encrypt/Decrypt

I agree the "Set Unspecified Request Values CMM" is valid and possible, but it feels like a really heavyweight solution. It would be a lot simpler to say that CMMs require the algorithm suite to be set in all cases.

Reply:
After offline discussion I'm at least convinced this is a reasonable solution - I'll make suggestions in the text for how we can justify it a bit more strongly.